src/server/twemcache-rs/src/protocol/data.rs

```diff
@@ -0,0 +1,425 @@
+        let data: &[u8] = self.data.borrow();
+        &data[self.key_index.0..self.key_index.1]
+    }
+
+    pub fn noreply(&self) -> bool {
+        self.noreply
+    }
+}
+
+#[derive(PartialEq, Eq, Debug)]
+pub struct SetRequest {
+    data: BytesMut,
+    key_index: (usize, usize),
+    expiry_index: (usize, usize),
+    flags: u32,
+    noreply: bool,
+    value_index: (usize, usize),
+}
+
+impl SetRequest {
+    pub fn key(&self) -> &[u8] {
+        let data: &[u8] = self.data.borrow();
+        &data[self.key_index.0..self.key_index.1]
+    }
+
+    pub fn value(&self) -> &[u8] {
+        let data: &[u8] = self.data.borrow();
+        &data[self.value_index.0..self.value_index.1]
+    }
+
+    pub fn flags(&self) -> u32 {
+        self.flags
+    }
+
+    pub fn noreply(&self) -> bool {
+        self.noreply
+    }
+}
+
+#[derive(PartialEq, Eq, Debug)]
+pub struct AddRequest {
+    data: BytesMut,
+    key_index: (usize, usize),
+    expiry_index: (usize, usize),
+    flags: u32,
+    noreply: bool,
+    value_index: (usize, usize),
+}
+
+impl AddRequest {
+    pub fn key(&self) -> &[u8] {
+        let data: &[u8] = self.data.borrow();
+        &data[self.key_index.0..self.key_index.1]
+    }
+
+    pub fn value(&self) -> &[u8] {
+        let data: &[u8] = self.data.borrow();
+        &data[self.value_index.0..self.value_index.1]
+    }
+
+    pub fn flags(&self) -> u32 {
+        self.flags
+    }
+
+    pub fn noreply(&self) -> bool {
+        self.noreply
+    }
+}
+
+#[derive(PartialEq, Eq, Debug)]
+pub struct ReplaceRequest {
+    data: BytesMut,
+    key_index: (usize, usize),
+    expiry_index: (usize, usize),
+    flags: u32,
+    noreply: bool,
+    value_index: (usize, usize),
+}
+
+impl ReplaceRequest {
+    pub fn key(&self) -> &[u8] {
+        let data: &[u8] = self.data.borrow();
+        &data[self.key_index.0..self.key_index.1]
+    }
+
+    pub fn value(&self) -> &[u8] {
+        let data: &[u8] = self.data.borrow();
+        &data[self.value_index.0..self.value_index.1]
+    }
+
+    pub fn flags(&self) -> u32 {
+        self.flags
+    }
+
+    pub fn noreply(&self) -> bool {
+        self.noreply
+    }
+}
+
+// TODO(bmartin): consider splitting the parsing functions up, cognitive
+// complexity is a little high for this function. Currently this way to re-use
+// the window iterators and avoid re-parsing the buffer once into command
+// specific parsing.
+pub fn parse(buffer: &mut BytesMut) -> Result<Request, ParseError> {
+    // no-copy borrow as a slice
+    let buf: &[u8] = (*buffer).borrow();
+
+    // check if we got a CRLF
+    let mut double_byte_windows = buf.windows(2);
+    if let Some(command_end) = double_byte_windows.position(|w| w == b"\r\n") {
+        // single-byte windowing to find spaces
+        let mut single_byte_windows = buf.windows(1);
+        if let Some(command_verb_end) = single_byte_windows.position(|w| w == b" ") {
+            let command_verb = &buf[0..command_verb_end];
```

Comment:
It might make sense to store the command_verb as an enum here to avoid cloning the ascii representation later.